argus/action_manager/windows.py

```diff
@@ -115,34 +117,95 @@ def get_installation_script(self):
         self.download_resource("windows/installCBinit.ps1",
                                r"C:\installCBinit.ps1")
 
+    def _execute(self, cmd, count=util.RETRY_COUNT, delay=util.RETRY_DELAY,
+                 command_type=util.CMD):
+        """Execute until success and return only the standard output."""
+        stdout, _, _ = self._client.run_command_with_retry(
+            cmd, count=count, delay=delay, command_type=command_type)
+        return stdout
+
+    def check_cbinit_installation(self):
+        """Check if Cloudbase-Init was installed successfully."""
+        LOG.info("Checking Cloudbase-Init installation.")
+
+        try:
+            python_dir = introspection.get_python_dir(
+                lambda cmd, command_type: self._client.run_command_with_retry(
+                    cmd, count=util.RETRY_COUNT,
+                    delay=util.RETRY_DELAY, command_type=command_type)[0]
+            )
+        except exceptions.ArgusError as exc:
+            LOG.warning("Could not check Cloudbase-Init installation: %s", exc)
+            self.cbinit_cleanup()
+            return False
+
+        check_cmd = r'& "{}\python.exe" -c "import cloudbaseinit"'.format(
+            python_dir)
+        try:
+            self._client.run_remote_cmd(
+                cmd=check_cmd, command_type=util.POWERSHELL)
+        except exceptions.ArgusError as exc:
+            LOG.debug("Cloudbase-Init installation failed: %s", exc)
+            self.cbinit_cleanup()
+            return False
+
+        LOG.info("Cloudbase-Init was successfully installed!")
+        return True
+
+    def cbinit_cleanup(self):
+        """Clean-up Cloudbase-Init if it's installed."""
+        LOG.info("Cleaning up Cloudbase-Init.")
+
+        try:
+            cbinit_dir = introspection.get_cbinit_dir(
+                lambda cmd, command_type: self._client.run_command_with_retry(
+                    cmd, count=util.RETRY_COUNT,
+                    delay=util.RETRY_DELAY, command_type=command_type)[0]
+            )
+        except exceptions.ArgusError as exc:
+            LOG.warning("Could not find Cloudbase-Init directory: %s", exc)
+        else:
+            self.rmdir(ntpath.dirname(cbinit_dir))
+
     def install_cbinit(self):
-        """Run the installation script for CloudbaseInit."""
-        LOG.debug("Installing Cloudbase-Init ...")
+        """Install Cloudbase-Init on the underlying instance."""
+        LOG.info("Trying to install Cloudbase-Init ...")
+
+        for _ in range(util.RETRY_COUNT):
+            for install_method in (self._run_installation_script,
+                                   self._deploy_using_scheduled_task):
+                try:
+                    install_method()
```

Comment:
At the first iteration, `install_method` is `_run_installation_script`. At the second iteration, if `_run_installation_script` throws an exception, it will be `_deploy_using_scheduled_task`.